include/mbedtls/platform_util.h

```diff
@@ -200,6 +200,51 @@ struct tm *mbedtls_platform_gmtime_r( const mbedtls_time_t *tt,
                                       struct tm *tm_buf );
 #endif /* MBEDTLS_HAVE_TIME_DATE */
 
+/**
+ * \brief Write data into buffer cache.
+ *
+ * \param[in] context   contexts pointer used for buffer management.
+ * \param[in] buf       buffer that will hold the data
+ * \param[in] len       how many bytes must be written
+ *
+ * \return         The (positive) number of bytes write if successful.
+ *
+ * \note           An alternative implement can put its context pointer in
+ *                 context.
+ *
+ * \note           For time being, it is design for TLS 1.3 early data.
+ *
+ * \note           Before mbedtls_ssl_read_early_data() is called,
+ *                 an intermediate buffer is needed to reserve 0-RTT data. And
+ *                 it is difficult to guarantee that data will be read
+ *                 immediately. The buffer cache is to reserve 0-RTT data until
+ *                 it is read.
+ *
+ * \note           If MBEDTLS_PLATFORM_BUF_CACHE_ALT is defined, then Mbed TLS
+ *                 will unconditionally use the alternative implementation for
+ *                 mbedtls_platform_buf_cache_{write,read}() supplied by the user
+ *                 at compile time.
+ */
+MBEDTLS_CHECK_RETURN_CRITICAL
+int mbedtls_platform_buf_cache_write( void **context, const void *buf, size_t len );
```

Comment:
I do not think we need that. The idea is to call `mbedtls_ssl_read_early_data()` with a "fresh SSL context" for which the handshake has not been started yet. Then `mbedtls_ssl_read_early_data()` starts the handshake and if we receive early data during the handshake we copy them in the buffer that was passed into `mbedtls_ssl_read_early_data()`.

Reply:
Is it expected ?
```c
int mbedtls_ssl_read_early_data( ssl, buf, size)
{
   
    ssl_conf_early_data_buf_point( ssl, buf, size ) ;

    while( handshake_is_not_over( ssl ) and "can read" )
    {
         ret = mbedtls_ssl_handshake_step();
         if( early_data is available)
         {
               return ( actually data size);
         }
    }
    return ”can not read";
}
```

Reply:
One more question.
When early data is configured enabled and user call `mbedtls_ssl_handshake` instead of `mbedtls_ssl_read_early_data`,  what action should be done? I suggest disable early data in this case.

Reply:
> Is it expected ?                                                              
>                                                                               
> ```c                                                                          
> int mbedtls_ssl_read_early_data( ssl, buf, size)                              
> {                                                                             
>                                                                               
>     ssl_conf_early_data_buf_point( ssl, buf, size ) ;                         
>                                                                               
>     while( handshake_is_not_over( ssl ) and "can read" )                      
>     {                                                                         
>          ret = mbedtls_ssl_handshake_step();                                  
>          if( early_data is available)                                         
>          {                                                                    
>                return ( actually data size);                                  
>          }                                                                    
>     }                                                                         
>     return ”can not read";                                                    
> }                                                                             
> ```                                                                           
                                                                                
I was even thinking at something like the following (not completely sure it can work like that though) where the reading of the early data is done at the level of `mbedtls_ssl_read_early_data()`, not inside `mbedtls_ssl_handshake()`.
                                                                              
```c                                                                            
int mbedtls_ssl_read_early_data( ssl, buf, size )                               
{                                                                               
    do
    {    
        /* Case of successive calls to `mbedtls_ssl_read_early_data()` to read early data */
        if( "early data available" )                                                 
        {                                                                           
            ret = ssl_read_real( ssl, buf, size );                                  
            if( ret < 0 )
                return( ret );     
            if( ret >= size )
                return( ret );                                                 
        }                                                                       
        
        ret = mbedtls_ssl_handshake();                                    
        if( ! "can read early data" )
        {                                           
            if( ret < 0 )
                return( ret );
         }
    } while(  "can read early data" );
    
    return ”can not read";                                                      
}                                                                               
```
Anyway it is not in scope of this PR anymore I think. We will discuss that further in the following PRs.

Reply:
> One more question. When early data is configured enabled and user call `mbedtls_ssl_handshake` instead of `mbedtls_ssl_read_early_data`, what action should be done? I suggest disable early data in this case.

I am not completely sure but it makes sense to me. The client is expected to send its early data along its ClientHello thus it is likely that after the server has processed the ClientHello and sent its ServerHello it has already received the first early data. And if we are not "inside `mbedtls_ssl_read_early_data()` we are not able to store the early data. We could probably return a new error code like MBEDTLS_ERR_SSL_WANT_READ_EARLY_DATA at that point but to start with let's do the simplest.

Reply:
I have split #6541 , Now #6541  only includes `mbedtls_ssl_read_early_data` . I will continue this thread in #6541.